Update bt_smarthub component making it compatible with smarthub 2

leroyshirto opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
version `0.2.0` of the `btsmarthub_devicelist` package makes it compatible with BT's home hub 2 routers. The API of the library has changed. This change makes the component code compatible with the changes to the library.

This change resolves the 404 errors members in the comunity have been experiencing
https://github.com/jxwolstenholme/btsmarthub_devicelist/issues/3
https://community.home-assistant.io/t/has-anyone-had-any-success-using-the-bt-home-hub-5-device-tracker-component-recently/38289/25

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
device_tracker:
  - platform: bt_smarthub
```
Note: testing this requires a BT Smarthub or Smarthub 2.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
PR for documentation change https://github.com/home-assistant/home-assistant.io/pull/11972

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

leroyshirto committed:
0.2.0 of the btsmarthub_devicelist package makes it compatable with BT's home hub 2.
The API has changed in the new version so this change also makes the component code compatible with the changes to the library.

homeassistant commented:
Hi @leroyshirto, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31292).

Once you do that we will be able to review and accept this pull request.

Thanks!

leroyshirto commented:
@jxwolstenholme I have implemented the changes you suggested in your feedback. Please could you review the technical implementation?

As we are now offering a config option to users I will look at updating the docs to reflect.

jxwolstenholme commented:
Looks good. Yes, agreed that the docs should be updated to confirm that Smarthub 2 support is now implemented.

jxwolstenholme reviewed (approved):
Approved from my end.

frenck commented on homeassistant/components/bt_smarthub/device_tracker.py:
```diff
@@ -15,9 +15,13 @@
-    {vol.Optional(CONF_HOST, default=CONF_DEFAULT_IP): cv.string}
+    {
+        vol.Optional(CONF_HOST, default=CONF_DEFAULT_IP): cv.string,
+        vol.Optional(CONF_SMARTHUB_MODEL, default=None): vol.In([None, 1, 2]),
```
```suggestion
        vol.Optional(CONF_SMARTHUB_MODEL): vol.In([1, 2]),
```

frenck commented on homeassistant/components/bt_smarthub/device_tracker.py:
```diff
@@ -35,8 +39,12 @@ def __init__(self, config):
         """Initialise the scanner."""
         _LOGGER.debug("Initialising BT Smart Hub")
         self.host = config[CONF_HOST]
+        self.smarthub_model = config[CONF_SMARTHUB_MODEL]
```
```suggestion
        self.smarthub_model = config.get(CONF_SMARTHUB_MODEL)
```

frenck reviewed (commented):
Thanks, @leroyshirto, left 2 small comments.

leroyshirto committed:
Update homeassistant/components/bt_smarthub/device_tracker.py

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

leroyshirto committed:
Update homeassistant/components/bt_smarthub/device_tracker.py

Co-Authored-By: Franck Nijhof <frenck@frenck.nl>

MartinHjelmare commented on homeassistant/components/bt_smarthub/device_tracker.py:
```diff
@@ -35,8 +39,12 @@ def __init__(self, config):
+        self.smarthub_model = config.get(CONF_SMARTHUB_MODEL)
         self.last_results = {}
         self.success_init = False
+        self.smarthub = BTSmartHub(
```
Please instantiate the `BTSmartHub` inside `get_scanner` and pass the instance to `BTSmartHubScanner` when instantiating that.

leroyshirto replied:
Please see my latest commit which addresses this request.

MartinHjelmare commented on homeassistant/components/bt_smarthub/device_tracker.py:
```diff
@@ -35,8 +39,12 @@ def __init__(self, config):
         """Initialise the scanner."""
         _LOGGER.debug("Initialising BT Smart Hub")
         self.host = config[CONF_HOST]
+        self.smarthub_model = config.get(CONF_SMARTHUB_MODEL)
```
This isn't needed here. It's just needed when creating the `BTSmartHub` instance.

leroyshirto replied:
Please see my latest commit which addresses this request.

MartinHjelmare reviewed (changes requested):
Please rebase on latest dev branch and solve the merge conflict.

leroyshirto committed:
Remove dep on config in BTSmartHubScanner

This should make BTSmartHubScanner easier to test as you can pass in a mock smarthub_client

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
/AzurePipelines run

MartinHjelmare commented:
The CI checks on this page aren't updated properly. The formatting check failed:
https://dev.azure.com/home-assistant/Core/_build/results?buildId=37555&view=logs&j=2fa2c639-d05e-530d-034e-4fb413fcea4e&t=88bd70b9-15d8-56b4-acb9-94f6fcce0e9d

Please run black and commit the changes.
```
black --fast homeassistant
```

leroyshirto committed:
Black format bt_smarthub

leroyshirto commented:
> The CI checks on this page aren't updated properly. The formatting check failed:
> https://dev.azure.com/home-assistant/Core/_build/results?buildId=37555&view=logs&j=2fa2c639-d05e-530d-034e-4fb413fcea4e&t=88bd70b9-15d8-56b4-acb9-94f6fcce0e9d
> 
> Please run black and commit the changes.
> 
> ```
> black --fast homeassistant
> ```

Do we need to re-run Azure Pipelines after I fixed the formatting or does that happen automatically?

MartinHjelmare commented:
A checks have passed now. :+1:

MartinHjelmare reviewed (approved):
Thanks!
